File: kotlin/Day09.kt
@file:Suppress("PackageDirectoryMismatch", "PackageName")

package Day09

import readInput


fun part1(input: List<String>): Int {
    var risk = 0
    for (i in input.indices) {
        for (j in input[i].indices) {
            if ((i == 0 || input[i][j] < input[i - 1][j]) &&
                    (i == input.size - 1 || input[i][j] < input[i + 1][j]) &&
                    (j == 0 || input[i][j] < input[i][j - 1]) &&
                    (j == input[i].length - 1 || input[i][j] < input[i][j + 1])) {
                risk += 1 + input[i][j].digitToInt()
            }
        }
    }
    return risk
}

fun part2(input: List<String>): Int {
    val cave = Array(input.size) { Array(input[0].length) { 0 } }

    for (i in input.indices) {
        for (j in input[i].indices) {
            cave[i][j] = input[i][j].digitToInt()
        }
    }

    val basins = mutableListOf<Int>()
    for (i in cave.indices) {
        for (j in cave[i].indices) {
            if (cave.isLowest(i, j)) {
                basins.add(cave.fillBasin(i, j))
            }
        }
    }

    return basins.sorted().takeLast(3).reduce { acc, i -> acc * i }
}

private fun Array<Array<Int>>.fillBasin(i: Int, j: Int): Int {
    if (this.isOutOfBounds(i, j) || this[i][j] == 9) return 0
    this[i][j] = 9
    return 1 + fillBasin(i + 1, j) + fillBasin(i - 1, j) + fillBasin(i, j + 1) + fillBasin(i, j - 1)
}

private fun Array<Array<Int>>.isOutOfBounds(i: Int, j: Int): Boolean {
    return (i == -1 || i == this.size || j == -1 || j == this[i].size)
}

private fun Array<Array<Int>>.isLowest(i: Int, j: Int): Boolean = (i == 0 || this[i][j] < this[i - 1][j]) &&
        (i == this.size - 1 || this[i][j] < this[i + 1][j]) &&
        (j == 0 || this[i][j] < this[i][j - 1]) &&
        (j == this[i].size - 1 || this[i][j] < this[i][j + 1])

fun main() {

    // test if implementation meets criteria from the description, like:
    val testInput = readInput("Day09_test")
    check(part1(testInput) == 15)
    check(part2(testInput) == 1134)

    val input = readInput("Day09")
    println(part1(input))
    println(part2(input))
}

File: swift/Day09.swift
import Foundation

@main
struct Day02 {
    
    static func part1(input: Array<String>) -> Int {
        var risk = 0
        for (i, _) in input.enumerated() {
            for (j, _) in input[i].enumerated() {
                if (i == 0 || input[i][j] < input[i - 1][j]) &&
                        (i == input.count - 1 || input[i][j] < input[i + 1][j]) &&
                        (j == 0 || input[i][j] < input[i][j - 1]) &&
                        (j == input[i].count - 1 || input[i][j] < input[i][j + 1]) {
                    risk += 1 + Int(String(input[i][j]))!
                }
            }
        }
        return risk
    }
    
    static func part2(input: Array<String>) -> Int {
        var cave = Array<Array<Int>>(repeating: Array(repeating: 0, count: input[0].count), count: input.count)
        
        for (i, _) in input.enumerated() {
            for (j, _) in input[i].enumerated() {
                cave[i][j] = Int(String(input[i][j]))!
            }
        }
        
        var basins = Array<Int>()
        for i in 0..<cave.count {
            for j in 0..<cave[i].count {
                if cave.isLowest(i: i, j: j) {
                    basins.append(cave.fillBasin(i, j))
                }
            }
        }
        
        return basins.sorted().suffix(3).reduce(1) { (acc, value) in acc * value }
    }
    
    static func main() {
        let testInput = Utils.readInput(name: "Day09_test")
        assert(part1(input: testInput) == 15)
        assert(part2(input: testInput) == 1134)
        
        let input = Utils.readInput(name: "Day09")
        print(part1(input: input), terminator: "\n")
        print(part2(input: input), terminator: "\n")
    }
}

private extension Array where Element == Array<Int> {
    func isLowest(i: Int, j: Int) -> Bool {
        return (i == 0 || self[i][j] < self[i - 1][j]) &&
        (i == self.count - 1 || self[i][j] < self[i + 1][j]) &&
        (j == 0 || self[i][j] < self[i][j - 1]) &&
        (j == self[i].count - 1 || self[i][j] < self[i][j + 1])
    }
    
    mutating func fillBasin(_ i: Int, _ j: Int) -> Int {
        if self.isOutOfBounds(i, j) || self[i][j] == 9 {
            return 0
        }
        self[i][j] = 9
        return 1 + fillBasin(i + 1, j) + fillBasin(i - 1, j) + fillBasin(i, j + 1) + fillBasin(i, j - 1)
    }
    
    func isOutOfBounds(_ i: Int, _ j: Int) -> Bool {
        return (i == -1 || i == self.count || j == -1 || j == self[i].count)
    }
}
